sympy/physics/quantum/tests/test_qapply.py

```diff
@@ -148,3 +153,354 @@ def test_issue24158_ket_times_op():
+                            -sin(2 * pi / n)**2 * Qubit(0) + 2 * sin(2 * pi / n) * cos(2 * pi / n) * Qubit(1) + \
+                            cos(2 * pi / n)**2 * Qubit(0)
+
+    m = 1
+    P2b = qapply(U ** m * Qubit(0), mul=False)
+    Ptsb = P2b.trigsimp() # fails to simplify to cos(m*2*pi/n)*|0>+sin(m*2*pi/n)*|1> for m == 2
+    assert Ptsb == (cos(m * 2 * pi / n) * Qubit(0) + sin(m * 2 * pi / n) * Qubit(1)).trigsimp() #fails for m >= 2
+
+    U = UGate(0, ImmutableMatrix([[E ** (2 * pi * I / n), 0], [0, E ** -(2 * pi * I / n)]]))
+    m = 5
+    assert qapply(U ** m) == U ** m # qapply relies on Pow, pow etc. to compute powers!
+    assert qapply(U ** m * Qubit(0)) == E ** (m * 2 * I * pi / n) * Qubit(0)
+
+    # simple addition of exponents
+    assert qapply(TGate(0) ** (m - n) * IdentityOperator(2) * TGate(0) ** n) == \
+        TGate(0) ** m # qapply relies on Pow for powers!
+    assert qapply(TGate(0) ** (m - n) * IdentityOperator(2) * TGate(0) ** n * Qubit(1)) == \
+        E ** -((8-m) * 2 * I * pi / 8) * Qubit(1)
+
+    # Dagger of unitary matrices U is U**-1, test qapply behaviour
+    U12 = UGate(0, ImmutableMatrix([[2, 0], [0, 2]]))
+    assert qapply(QubitBra(1)*U12) == QubitBra(1)*U12 # As long as there is no method defined for Bra*UGate
+    assert qapply(Dagger(U12)*Qubit(1), dagger=True) == U12**(-1) * Qubit(1) # (Dagger(UGate)==UGate**-1) which doesn't know about |1>
+    assert Dagger(Dagger(U12)) == U12
+    assert qapply(QubitBra(1)*Dagger(U12), dagger=True) == 2 * QubitBra(1) # with dagger=True it boils down to Dagger(U12*Qubit(1))
+    assert qapply(Dagger(Dagger(U12))*Dagger(QubitBra(1))) == 2 * Qubit(1)
+
+
+def test_powers2022_11():
+    from sympy import exp, I, pi
+    from sympy.matrices import ImmutableMatrix, eye
+
+    # Use an extend definition of UGate*UGate with diverse result types in order
+    # to build test cases with commutative factors etc.
+    def UGate_apply_operator_UGate(u1,u2, **options):
+        if u1.args[0] != u2.args[0]: # This case is not implemented
+            raise NotImplementedError("UGate*UGate on different target qubits")
+        u1id = (u1.args[1] == u1.args[1][0,0] * eye(u1.args[1].shape[0]))
+        u2id = (u2.args[1] == u2.args[1][0,0] * eye(u2.args[1].shape[0]))
+        if u1id and u2id:
+            return u1.args[1][0,0] * u2.args[1][0,0] # scalar result
+        elif u1id:
+            return u1.args[1][0,0] * u2 # scalar * UGate
+        elif u2id:
+            return u2.args[1][0,0] * u1 # scalar * UGate
+        else:
+            res = u1.args[1] * u2.args[1] # assuming res again immutableMatrix
+            if res == res[0,0] * eye(res.shape[0]):
+                return res[0,0]  # scalar result
+            else:
+                return UGate(u1.args[0], res)
+    prev_op = getattr(UGate, "_apply_operator_UGate", None)
+    setattr(UGate, "_apply_operator_UGate", UGate_apply_operator_UGate)
```

Comment:
I don't quite understand what this is doing but I guess it should use `try/finally` to change this attribute.

Reply:
getattr essentially has try/finally built in, and will never fail.
setattr the same: setting an attribute on a writable object cannot fail. 
Moreover this is for creating a test bed only.

Reply:
> Moreover this is for creating a test bed only.

If the test fails because of an exception after this call to `setattr` then the attribute should be restored so that it does not affect other tests running in the rest of the test suite.

Code within any one test function should avoid altering global state (which also would affect running tests in parallel in different threads). If it is really necessary to alter global state then that state needs to be restored even in the event of a test failure.

Reply:
Fixed completely by using a local clone of the global class.